client/components/infinite-scroll/index.js

```diff
@@ -3,36 +3,118 @@
 /**
  * External dependencies
  */
-import { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
-import { throttle } from 'lodash';
+import { throttle, defer } from 'lodash';
+import afterLayoutFlush from 'lib/after-layout-flush';
 
 const SCROLL_CHECK_RATE_IN_MS = 400;
 
-export default class InfiniteScroll extends Component {
-	static propTypes = {
-		nextPageMethod: PropTypes.func.isRequired,
+const hasIntersectionObserver = typeof window !== 'undefined' && 'IntersectionObserver' in window;
+
+const propTypeDefinition = {
+	nextPageMethod: PropTypes.func.isRequired,
+};
+
+class InfiniteScrollWithIntersectionObserver extends React.Component {
+	static propTypes = propTypeDefinition;
+
+	observedElement = React.createRef();
+
+	hasScrolledPastBottom = false;
+
+	componentDidMount() {
+		if ( this.observedElement.current ) {
+			this.observer = new IntersectionObserver( this.handleIntersection, {
+				rootMargin: '100%',
+				threshold: 1.0,
+			} );
+			this.observer.observe( this.observedElement.current );
+		}
+	}
+
+	componentWillUnmount() {
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
+		clearTimeout( this.deferredPageFetch );
+	}
+
+	getNextPage() {
+		this.props.nextPageMethod( { triggeredByScroll: this.hasScrolledPastBottom } );
+	}
+
+	componentDidUpdate() {
+		if ( ! this.deferredPageFetch && ! this.hasScrolledPastBottom ) {
+			// We still need more pages, so schedule a page fetch.
+			this.deferredPageFetch = defer( () => {
+				if ( ! this.hasScrolledPastBottom ) {
+					this.getNextPage();
+				}
+				this.deferredPageFetch = null;
+			} );
+		}
+	}
+
+	handleIntersection = entries => {
+		if ( ! entries || ! entries[ 0 ] ) {
+			return;
+		}
+
+		if ( entries[ 0 ].isIntersecting ) {
+			this.getNextPage();
+		} else {
+			// The observed element is no longer in view, so future changes must
+			// be caused by scrolling.
+			this.hasScrolledPastBottom = true;
+		}
 	};
 
-	checkScrollPositionHandler = throttle(
-		() => this.checkScrollPosition( true ),
-		SCROLL_CHECK_RATE_IN_MS
-	);
+	render() {
+		return <div ref={ this.observedElement } />;
+	}
+}
+
+class InfiniteScrollWithScrollEvent extends React.Component {
+	static propTypes = propTypeDefinition;
 
 	componentDidMount() {
 		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
```

Comment:
The `checkScrollPositionHandler` expect a `triggeredByScroll` param that is `true`. In this case, it will receive the scroll even as parameter, which is also a truthy value, but still...

Reply:
You're totally right, it's ugly. I created a new instance method for event handling, that always passes `true`.